File: Editing/Editing View/Workspace/PhotoEditingCanvasView.m
//  Created by Geoff Pado on 8/27/19.
//  Copyright © 2019 Cocoatype, LLC. All rights reserved.

#import "PhotoEditingCanvasView.h"
#import "PhotoEditingCanvasBrushStrokeView.h"

#define STANDARD_LINE_WIDTH 10.0

@implementation PhotoEditingCanvasView

- (instancetype)init
{
    if ((self = [super initWithFrame:CGRectZero])) {
        self.accessibilityIgnoresInvertColors = YES;
        self.drawingPolicy = PKCanvasViewDrawingPolicyAnyInput;
        self.backgroundColor = [UIColor clearColor];
        self.color = [UIColor blackColor];
        self.opaque = NO;
        self.overrideUserInterfaceStyle = UIUserInterfaceStyleLight;
        self.tool = [self toolForZoomScale:1.0];
        self.translatesAutoresizingMaskIntoConstraints = NO;
    }

    return self;
}

- (void)setColor:(UIColor *)color {
    _color = color;
    PKInkingTool *currentTool = (PKInkingTool *)[self tool];
    PKInkingTool *newTool = [[PKInkingTool alloc] initWithInkType:currentTool.inkType color:color width:currentTool.width];
    self.tool = newTool;
}

// MARK: Tool Creation

- (CGFloat)currentLineWidth {
    return [((PKInkingTool *)[self tool]) width];
}

- (PKTool *)toolForZoomScale:(CGFloat)zoomScale {
    return [[PKInkingTool alloc] initWithInkType:PKInkTypeMarker color:self.color width:[self adjustedLineWidthForZoomScale:zoomScale]];
}

// MARK: Zoom Handling

- (void)updateToolWithCurrentZoomScale:(CGFloat)currentZoomScale {
    self.tool = [self toolForZoomScale:currentZoomScale];
}

- (CGFloat)adjustedLineWidthForZoomScale:(CGFloat)zoomScale {
    return STANDARD_LINE_WIDTH * pow(zoomScale, -1.0);
}

// MARK: Touch Handling

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesBegan:touches withEvent:event];
    [self.brushStrokeView touchesBegan:touches withEvent:event];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesMoved:touches withEvent:event];
    [self.brushStrokeView touchesMoved:touches withEvent:event];
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesCancelled:touches withEvent:event];
    [self.brushStrokeView touchesCancelled:touches withEvent:event];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesEnded:touches withEvent:event];
    [self.brushStrokeView touchesEnded:touches withEvent:event];
}

- (PhotoEditingCanvasBrushStrokeView *)brushStrokeView {
    return (PhotoEditingCanvasBrushStrokeView *)[self superview];
}

@end

File: Canvas/PhotoEditingCanvasView.swift
//  Created by Geoff Pado on 7/31/19.
//  Copyright © 2019 Cocoatype, LLC. All rights reserved.

import PencilKit
import UIKit

@available(iOS 13.0, *)
class PhotoEditingCanvasView: PKCanvasView {
    init() {
        super.init(frame: .zero)
        allowsFingerDrawing = true
        backgroundColor = .clear
        isOpaque = false
        tool = PKInkingTool(.marker, color: .black, width: PhotoEditingCanvasView.standardLineWidth)
        translatesAutoresizingMaskIntoConstraints = false
    }

    var currentLineWidth: CGFloat {
        (tool as? PKInkingTool)?.width ?? PhotoEditingCanvasView.standardLineWidth
    }

    func updateTool(currentZoomScale: CGFloat) {
        tool = PKInkingTool(.marker, color: .black, width: adjustedLineWidth(forZoomScale: currentZoomScale))
    }

    private func adjustedLineWidth(forZoomScale zoomScale: CGFloat) -> CGFloat {
        return PhotoEditingCanvasView.standardLineWidth * pow(zoomScale, -1.0)
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        brushStrokeView?.touchesBegan(touches, with: event)
    }

    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesMoved(touches, with: event)
        brushStrokeView?.touchesMoved(touches, with: event)
    }

    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        brushStrokeView?.touchesEnded(touches, with: event)
    }

    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesCancelled(touches, with: event)
        brushStrokeView?.touchesCancelled(touches, with: event)
    }

    // MARK: Boilerplate

    private static let standardLineWidth = CGFloat(10.0)

    private var brushStrokeView: PhotoEditingCanvasBrushStrokeView? {
        return (superview as? PhotoEditingCanvasBrushStrokeView)
    }

    @available(*, unavailable)
    required init(coder: NSCoder) {
        let className = String(describing: type(of: self))
        fatalError("\(className) does not implement init(coder:)")
    }
}
